packages/protocol/contracts/common/B12.sol

```diff
@@ -0,0 +1,898 @@
+    return Fp2(res0, res1);
+  }
+
+  function fpNormal2(Fp memory a, uint256 idx) internal view returns (Fp memory) {
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
+    return fpModExp2(a, idx, 1, p);
+  }
+
+  function fpNormal(Fp memory a) internal view returns (Fp memory) {
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
+    return fpModExp(a, 1, p);
+  }
+
+  function mapToG2(Fp2 memory x, Fp2 memory hint1, Fp2 memory hint2, bool greatest)
+    internal
+    view
+    returns (G2Point memory)
+  {
+    Fp2 memory one = Fp2(Fp(0, 1), Fp(0, 0));
+    Fp2 memory res = fp2Add(fp2Mul(x, fp2Mul(x, x)), one);
+    Fp2 memory sqhint1 = fp2Mul(hint1, hint1);
+    Fp2 memory sqhint2 = fp2Mul(hint2, hint2);
+    require(Fp2Eq(sqhint1, res), "y1 not sqrt");
+    require(Fp2Eq(sqhint2, res), "y2 not sqrt");
+    require(fp2Gt(hint1, hint2), "y1 not greatest");
+    G2Point memory p = G2Point(x, greatest ? hint1 : hint2);
+    return p;
+  }
+
+  function mapToG1(Fp memory x, Fp memory hint1, Fp memory hint2, bool greatest)
+    internal
+    view
+    returns (G1Point memory)
+  {
+    Fp memory base = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
+    Fp memory one = Fp(0, 1);
+    Fp memory res = fpAdd(fpModExp(x, 3, base), one);
+    Fp memory sqhint1 = fpModExp(hint1, 2, base);
+    Fp memory sqhint2 = fpModExp(hint2, 2, base);
+    require(FpEq(sqhint1, res), "y1 not sqrt");
+    require(FpEq(sqhint2, res), "y2 not sqrt");
+    require(fpGt(hint1, hint2), "y1 not greatest");
+    return G1Point(x, greatest ? hint1 : hint2);
+  }
+
+  function parsePointGen(bytes memory h, uint256 offset)
+    internal
+    pure
+    returns (uint256, uint256, uint256)
+  {
+    uint256 a = 0;
+    uint256 b = 0;
+    for (uint256 i = 0; i < 32; i++) {
+      uint256 byt = uint256(uint8(h[offset + i]));
+      b = b + (byt << (i * 8));
+    }
+    for (uint256 i = 0; i < 15; i++) {
+      uint256 byt = uint256(uint8(h[offset + i + 32]));
+      a = a + (byt << (i * 8));
+    }
+    return (a, b, uint256(uint8(h[offset + 47])));
+  }
+
+  function parsePoint(bytes memory h, uint256 offset) internal pure returns (Fp memory, bool) {
+    (uint256 a, uint256 b, uint256 byt) = parsePointGen(h, offset);
+    a = a + ((byt & 0x7f) << (15 * 8));
+    return (Fp(a, b), byt & 0xa0 != 0);
+  }
+
+  function parseSimplePoint(bytes memory h, uint256 offset) internal pure returns (Fp memory) {
+    Fp memory res = Fp(0, 0);
+    parseSimplePoint(h, offset, res);
+    return res;
+  }
+
+  function parseSimplePoint(bytes memory h, uint256 offset, Fp memory p) internal pure {
+    /* (uint256 a, uint256 b, uint256 byt) = parsePointGen(h, offset);
+    a = a + (byt << 15*8); */
```

Comment:
Not sure if these are explanatory comments or unused code. Seems more like the latter? If that's the case, do we know for sure which is correct, if it's safe to remove the commented out code?